_posts/got-the-time.md

```diff
@@ -0,0 +1,244 @@
+
+```csharp
+class OverdueInvoicePolicy :
+   Saga<OverdueInvoicePolicy.OverdueInvoiceData>,
+   IAmStartedByMessages<InvoiceIssued>,
+   IHandleMessagesMessages<InvoicePaid>,
+   IHandleTimeouts<CheckPayment>
+{
+   public async Task Timeout(CheckPayment state, IMessageHandlerContext context)
+   {
+      //...
+   }
+}
+```
+
+(rest of the code omitted for clarity)
+
+What are we going to do when the timeout expires? When the timeout is due, NServiceBus dispatches the `CheckPayment` timeout message to the saga; at this point, we can check the payment status of the invoice and act accordingly:
+
+```csharp
+public async Task Timeout(CheckPayment state, IMessageHandlerContext context)
+{
+   var invoiceNumber = Data.InvoiceNumber;
+   var isInvoicePaid = _invoiceService.IsInvoicePaid(invoiceNumber);
+   if(!isInvoicePaid)
+   {
+      await context.Publish(new InvoiceOverdueEvent()
+      {
+         InvoiceNumber = invoiceNumber
+      });
+   }
+
+   MarkAsComplete();
+}
+```
+
+> The invoice service (`_invoiceService`), injected into the saga via IoC/DI, is a component that, given an invoice number, returns its payment status.
+
+When time is due, the saga, using the provided invoice service, checks the payment status. If the invoice is overdue, it raises an event accordingly and then marks itself as complete; If the invoice is settled, it immediately marks itself as complete. The saga has one single responsibility, to check for overdue invoices; the saga is not responsible for any further action; it delegates other activities to the endpoints handling the `InvoiceOverdue` event.
+
+### Checking is redundant
+
+Do we need to use an invoice service to check if a given invoice is overdue or not? The answer is no, and this comes with a significant simplification of the saga code. In our scenario, all the messages that the saga handles correctly describe all the possible options. When customers order, the system issues the `InvoiceIssued` event; when they pay, the system publishes the `InvoicePaid` event. There is no need to check for the status when the `CheckPayment` timeout expires from the overdue invoices saga perspective. When an invoice is paid, the infrastructure deletes the corresponding saga (mark as complete); this makes it so that NServiceBus discards the timeout. When NServiceBus successfully dispatches the timeout to the saga, it means that the customer didn't pay that specific invoice yet. The saga can assume that if the timeout reaches the saga, it can publish the `InvoiceOverdue` event; no one paid the invoice yet.
+
+### Order matters: Issue, pay, mark as complete
+
+At this point, an interesting question is:
+
+> What happens to the timeout if the `InvoicePaid` message arrives before the timeout expires?
+
+Previously, we defined it as completed when the saga handles the `InvoicePaid` message. However, if the saga receives messages in the following order:
```

Comment:
I'm confused by this part. In the last section, you said that when an InvoicePaid event happens, the saga is marked as complete and the timeout is deleted. This section says when an invoice is marked as paid, the timeout will stick around.

Reply:
Good catch. The previous sentence is wrong, the timeout is not deleted. It will be when it expires and the saga doesn't exist anymore.